### PR TITLE
Implemented CompositionCustomVisual

### PR DESCRIPTION
**Pull request by kekekeks**
https://user-images.githubusercontent.com/1067584/207939567-01eed059-ef43-4a07-9955-b5f28c6f2b50.mp4
```cs
var ch = compositor.CreateCustomVisual(new MyHandler());
ElementComposition.SetElementChildVisual(control, ch);


public abstract class CompositionCustomVisualHandler
{
  // User-implementable
  public virtual void OnMessage(object message);
  public virtual void OnAnimationFrameUpdate();
  public abstract void OnRender(ImmediateDrawingContext drawingContext);
  public virtual Rect GetRenderBounds();
  // API
  protected void RegisterForNextAnimationFrameUpdate();
  protected void Invalidate();
  protected TimeSpan CompositionNow { get; };
  protected Vector2 EffectiveSize { get; }
}

public class CompositionCustomVisual
{
  public void SendHandlerMessage(object message);
}
```

By calling `RegisterForNextAnimationFrameUpdate` the handler asks the compositor to call `OnAnimationFrameUpdate`, which in turn may ask to `Invalidate` the visual and trigger its rendering.

This allows one to have custom-drawn animations or gif/video playback directly on the render thread without a roundtrip to the UI thread.

This API will replace `ICustomDrawOperation` eventually.

`ImmediateDrawingContext` is a `DrawingContext` clone that accepts only thread-safe objects as parameters. Since such visuals typically run on a non-UI thread, we should have a proper API that won't surprise the user with unexpected crashes. `IImmutableBrush` interface is introduced for the same purpose.

### CODE REVIEW

**Comment by robloo** [comment]
WinUI has a SpectrumBrush that renders the color picker spectrum using a Composition Brush (if I recall correctly). This is used on newer versions of the OS that support it for performance reasons.

Does what you do here enable that in Avalonia at all?

**Comment by maxkatz6** [comment]
@robloo UWP doesn't have exact alternative to this PR API, although it somewhat can be done with Win2D I assume. This API is more about having a `Render(Context)` method which can be used with compositor infrastructure on a render thread.

> WinUI has a SpectrumBrush that renders the color picker spectrum using a Composition Brush (if I recall correctly).
You mean, [XamlCompositionBrushBase](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.media.xamlcompositionbrushbase?view=winrt-22621)? There is no alternative yet.

But you should be able to use this PR API or old API of ICustomDrawOperation (which is being replaced by this PR) to implement lower-level spectrum rendering. It just won't be possible to use as a Brush (as we need a XamlCompositionBrushBase alternative first), instead it needs to be a visual.

**Comment by robloo** [comment]
Alright, I saw IImmutableBrush and the ability to render entirely on the composition thread and thought perhaps it could be done in a similar way as WinUI. Just a curiosity.

**Comment by kekekeks** [comment]
Composition brushes are planned, but probably won't make it to 11.0.

File: src/Avalonia.Base/Rendering/Composition/CompositionCustomVisualHandler.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        get
+        {
+            VerifyAccess();
+            return _host!.Compositor.ServerNow;
+        }
+    }
+
+    public virtual Rect GetRenderBounds() =>
+        new(0, 0, EffectiveSize.X, EffectiveSize.Y);
```
Comment (maxkatz6) [inline src/Avalonia.Base/Rendering/Composition/CompositionCustomVisualHandler.cs:50]:
```suggestion
        new(0, 0, _host!.Size.X, _host.Size.Y);
```

I think we can avoid VerifyAccess call here, don't we?

File: src/Avalonia.Base/Rendering/Composition/Server/ServerCustomCompositionVisual.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        for (var c = 0; c < count; c++)
+        {
+            try
+            {
+                _handler.OnMessage(reader.ReadObject()!);
+            }
+            catch (Exception e)
+            {
+                Logger.TryGet(LogEventLevel.Error, LogArea.Visual)
+                    ?.Log(_handler, $"Exception in {nameof(CompositionCustomVisualHandler)} {{0}}", e);
```
Comment (maxkatz6) [inline src/Avalonia.Base/Rendering/Composition/Server/ServerCustomCompositionVisual.cs:32]: I think we can log here type name of the handler. Would help a lot with diagnosing errors here.

**Review by maxkatz6 — APPROVED** [review]

**Review by maxkatz6 — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027717-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**